node/engine/chainservice/eth_chainservice.go

```diff
@@ -263,113 +265,102 @@ func (ecs *EthChainService) dispatchChainEvents(logs []ethTypes.Log) error {
 			if sErr != nil {
 				errorChan <- fmt.Errorf("subscribeFilterLogs failed on resubscribe: %w", err)
 				break out
 			}
-			ecs.logger.Trace().Msg("resubscribed to filtered logs")
+			ecs.logger.Trace().Msg("resubscribed to filtered event logs")
+
+		case err := <-newBlockSub.Err():
+			if err != nil {
+				errorChan <- fmt.Errorf("received error from new block subscription channel: %w", err)
+				break out
+			}
+
+			// If the error is nil then the subscription was closed and we need to re-subscribe.
+			// This is a workaround for https://github.com/ethereum/go-ethereum/issues/23845
+			var sErr error
+			newBlockSub, sErr = ecs.chain.SubscribeNewHead(ecs.ctx, newBlockChan)
+			if sErr != nil {
+				errorChan <- fmt.Errorf("subscribeNewHead failed on resubscribe: %w", err)
+				break out
+			}
+			ecs.logger.Trace().Msg("resubscribed to new blocks")
 
 		case <-time.After(RESUB_INTERVAL):
 			// Due to https://github.com/ethereum/go-ethereum/issues/23845 we can't rely on a long running subscription.
 			// We unsub here and recreate the subscription in the next iteration of the select.
-			sub.Unsubscribe()
-		case naChainEvent := <-logs:
-			// Make sure we care about this event. If we do, process asynchronously
+			eventSub.Unsubscribe()
+
+		case chainEvent := <-eventChan:
 			for _, topic := range topicsToWatch {
-				if naChainEvent.Topics[0] == topic {
-					go ecs.waitToProcessEvent(naChainEvent, errorChan)
-					break
+				if chainEvent.Topics[0] == topic {
+					ecs.logger.Debug().Msgf("queueing new chainEvent from block: %d", chainEvent.BlockNumber)
+					heap.Push(eventQueue, chainEvent)
 				}
 			}
-		}
-	}
-}
 
-func (ecs *EthChainService) waitToProcessEvent(chainEvent ethTypes.Log, errorChan chan<- error) {
-	ecs.logger.Debug().Msgf("event emitted in tx %s (topic: %s)", chainEvent.TxHash.Hex(), chainEvent.Topics[0])
-	initialBlock := big.NewInt(int64(chainEvent.BlockNumber))
-
-	ticker := time.NewTicker(BLOCK_QUERY_DELAY_SECONDS * time.Second)
-	defer ticker.Stop()
-
-	numTicks := 0
-
-	// Call this function once initially, then after each tick
-	processTick := func() error {
-		numTicks++
-		header, err := ecs.chain.HeaderByNumber(context.Background(), nil)
-		if err != nil {
-			return fmt.Errorf("error in HeaderByNumber: %w", err)
-		}
-
-		// TODO: check if the block containing the transaction is still part of the chain
-		confirmations := int(header.Number.Int64() - initialBlock.Int64())
-		ecs.logger.Trace().Msgf("event block confirmations: %d (goal: %d)", confirmations, REQUIRED_BLOCK_CONFIRMATIONS)
-
-		if confirmations >= REQUIRED_BLOCK_CONFIRMATIONS {
-			err := ecs.dispatchChainEvents([]ethTypes.Log{chainEvent})
-			if err != nil {
-				return fmt.Errorf("error in dispatchChainEvents: %w", err)
+		case newBlock := <-newBlockChan:
+			ecs.logger.Debug().Msgf("detected new block: %d", newBlock.Number.Uint64())
+			// When we get a new block, check the events at the top of the queue.
```

Comment:
It doesn't seem like this takes into account uncle blocks. Probably ok for now, but we should file an issue for future work.

Reply:
Added new issue here: https://github.com/statechannels/go-nitro/issues/1525

Reply:
I think with proof of stake [uncle blocks aren't a thing](https://ethereum.stackexchange.com/questions/139390/are-uncle-ommer-blocks-still-possible-after-the-merge).  However it probably still makes sense to consider this so we can run against chains where uncle blocks are a thing.